specs/switch-driver-capabilities.md

```diff
@@ -0,0 +1,54 @@
+
+# Expose network driver capabilites
+-----------
+
+[Issue 410](https://github.com/CCI-MOC/hil/issues/410)
+
+We want to expose switch driver capabilities. Capabilities could include:
+
+- Jumbo frames are per-switch on our Dell switches and per-port on our Ciscos (#755)
+-  QoS (#384)
+- Disabling STP (#779)
+- DHCP/ARP Snooping (#471)
+- Enabling the tagging of DHCP packets with the port number (from @okrieg or @pjd-nu)
+
+# Solution
+-----------
+
+* All switch drivers would have a method called `get_capabilities()`.
+
+* This method would return a list with all capabilities supported by the switch.
+Its implementation could be flexible; it could simply return a hardcoded list, or generate it from some database.
+
+
+* method `ensure_legal_operations()` will be updated to check for capabilities.
+Currently, this method is called by the API before anything is queued for the networking
```

Comment:
I think I might be wrong about this. Any other switch capability we add will be exposed via a different API. `ensure_legal_ops()` is only called by `node_connect/detach_network()`. So I can leave `ensure_legal_ops()` as it is. Any other API would just do `capability in switch.get_capabilities()`.

Reply:
Yeah, I think this bullet point probably isn't quite right.

Reply:
done. now that I think of it, the spec thing for this is a bit of an overkill. I think of adding at least one capability when introducing this (perhaps jumbo frames on the brocade).